### PR TITLE
Add `min_mask_region_area` to SAM `generate()` and scale crop point prompts to the resized crop

### PR DESCRIPTION
**Pull request by raimbekovm**
## summary

- add `min_mask_region_area` to `Predictor.generate()`: when > 0, the existing `remove_small_regions` runs on the generated masks (islands and holes below the threshold, in original-image pixels, then box NMS) and scores and boxes are re-synced; the default 0 leaves the output unchanged; `remove_small_regions` now returns the documented `(masks, keep)` pair on an empty input
- scale the crop point grid to the resized crop for `crop_n_layers >= 1`: the crop is interpolated to the model input while the points stayed in crop pixels, so each crop was prompted only in its top-left ~42%
- apply the predictor `conf` inside `generate()` after the cross-crop NMS, where postprocess applies it anyway, so direct `generate()` calls and `predictor(...)` agree and a mask postprocess would drop cannot decide the cleanup NMS
- read the source (height, width) through one `src_shape` property for HWC arrays and CHW tensor sources at every prompt and area site; one commit per change, and the segment-everything docs example shows the new argument

## measured

mobile_sam.pt, bus.jpg (1080x810), imgsz 1024, CPU unless noted

| run | before | after |
| --- | --- | --- |
| masks with `min_mask_region_area=20` / `100` | 83 / 83 (argument rejected) | 77 / 76 |
| masks carrying a component below 100 px after resize to the original image | 12 | 0 |
| `sam2.1_b.pt` on MPS, `min_mask_region_area=100` | 22 | 21 |
| zidane.jpg docs example (`crop_n_layers=1, points_stride=64`), `min_mask_region_area=100` | 48 | 41 |
| decoder point extent on layer-1 crops (crop resized to 1024 px) | 665 px | 992 px, same as layer 0 |
| masks with `crop_n_layers=1, points_stride=16` | 95 | 84 |
| `crop_n_layers=0` control | 48 | 48 |
| source shape read for a 1024x1024 tensor source | (3, 1024) | (1024, 1024) |
| cleanup cost for 83 masks at 1024² | - | 0.76 s CPU, 1.1 s MPS (generate: 21.5 s CPU) |

Default calls (`min_mask_region_area=0`, `crop_n_layers=0`) return the same 83 masks as before.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Added configurable small-region cleanup to SAM `generate()` and corrected crop prompt scaling for resized crops, while centralizing source-shape handling across predictors.

### 📊 Key Changes
- Added `min_mask_region_area`, which removes small disconnected mask regions and holes, re-runs NMS, and synchronizes scores and bounding boxes; the default `0` preserves the existing behavior.
- Applied the predictor confidence threshold inside `generate()` after cross-crop NMS so direct `generate()` results and predictor postprocessing use the same filtering stage.
- Scaled crop point grids to the resized model-input dimensions, ensuring prompts cover the full crop when `crop_n_layers >= 1`.
- Added the `src_shape` property for consistent height and width extraction from HWC image arrays and CHW tensor sources, and used it across prompt preparation and area calculations.
- Updated `remove_small_regions()` to return a `(masks, keep)` pair for empty inputs and documented the new argument in the SAM example.

### 🎯 Purpose & Impact
- SAM crop-based generation now samples prompts across the entire resized crop rather than only its upper-left portion.
- Callers can enable cleanup of small mask regions using an area threshold measured in original-image pixels; cleaned masks have corresponding scores and boxes updated.
- Default generation without cleanup and without crops remains unchanged.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the SAM generate changes, crop prompt scaling, small-region cleanup, score/box synchronization, empty-input handling, and documentation update. No concrete correctness, compatibility, security, or performance regressions found. LGTM.

File: ultralytics/models/sam/predict.py [hunk 1]
```diff
@@ -438,6 +441,15 @@ def generate(
             keep = torchvision.ops.nms(pred_bboxes, scores, crop_nms_thresh)
             pred_masks, pred_bboxes, pred_scores = pred_masks[keep], pred_bboxes[keep], pred_scores[keep]
 
+        if min_mask_region_area > 0:
+            idx = pred_scores > self.args.conf  # postprocess drops these, so they must not suppress masks it keeps
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/predict.py]: 📝 **LOW**: When `generate()` is called directly with a lower `conf_thres`, enabling cleanup applies a second filter using `self.args.conf`. For example, with the default predictor confidence of 0.25, `generate(im, conf_thres=0.1, min_mask_region_area=100)` silently drops masks scoring 0.1–0.25, although those masks are retained when cleanup is disabled and `conf_thres` is documented as the generate-time filter. Keep masks needed only for postprocess filtering out of cleanup NMS without removing them from the raw `generate()` result, or otherwise preserve the `conf_thres` contract.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Review 2: crop prompt scaling and mask/score/box synchronization remain correct, and no issues were found in the documentation change or empty-input handling. One bounded correctness issue remains: enabling cleanup can make direct `generate()` calls apply `self.args.conf` in addition to the documented `conf_thres`, dropping otherwise valid low-score masks.

<details><summary>💬 Posted 1 inline comment</summary>

- 📝 **LOW** `ultralytics/models/sam/predict.py:445` When `generate()` is called directly with a lower `conf_thres`, enabling cleanup applies a second filter using `self.args.conf`. For example, with the default predictor confidence of 0.25, `generate(im, conf_thres=0.1, min_mask_region_area=100)` silently drops masks scoring 0.1–0.25, although those masks are retained when cleanup is disabled and `conf_thres` is documented as the generate-time filter. Keep masks needed only for postprocess filtering out of cleanup NMS without removing them from…
</details>

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Review 3: crop point scaling, original-pixel area conversion, mask/score/box synchronization, empty-input handling, and documentation remain correct. The previously identified confidence-threshold mismatch remains unresolved; no additional issues were found.

<details><summary>💬 Posted 1 inline comment</summary>

- 📝 **LOW** `ultralytics/models/sam/predict.py:448` When this new option is used through `Predictor.__call__`, `conf_thres` is forwarded only to `generate()`, while `postprocess()` still filters the returned scores with `self.args.conf`. For example, `predictor(source=..., conf_thres=0.1, min_mask_region_area=100)` can discard masks scoring 0.1–0.25 despite the documented generate-time threshold accepting them. Route the effective confidence threshold consistently into postprocessing, or apply the final confidence filter inside `generate()` so t…
</details>

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 4

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 3, the confidence-threshold mismatch remains open, and this pass identifies a new tensor-input bug in original-pixel area scaling. Crop prompt scaling, mask/score/box synchronization, empty-input handling, and the documentation update otherwise look correct.

<details><summary>💬 Posted 2 inline comments</summary>

- 📝 **LOW** `ultralytics/models/sam/predict.py:445` For tensor sources, `self.batch[1]` is a BCHW tensor (`LoadTensor`), so `self.batch[1][0].shape[:2]` yields `(C, H)` rather than `(H, W)`. With a non-square tensor input and cleanup enabled, `gain` therefore converts the original-pixel threshold incorrectly; for example, a `3x640x960` input produces `gain=1.5` instead of `1`. Derive dimensions from the tensor's trailing axes while retaining the existing HWC handling for image lists.
- 📝 **LOW** `ultralytics/models/sam/predict.py:448` The confidence-threshold mismatch from the previous review remains unresolved: when called through `Predictor.__call__(..., conf_thres=0.1, min_mask_region_area=100)`, `conf_thres` filters candidates inside `generate()`, but `postprocess()` still applies `self.args.conf` (normally 0.25). Masks scoring 0.1–0.25 are therefore discarded after this new cleanup path, contrary to the generate-time threshold supplied by the caller. Pass the effective threshold through postprocessing or apply the final…
</details>

File: ultralytics/models/sam/predict.py [hunk 2]
```diff
@@ -382,7 +385,7 @@ def generate(
             x1, y1, x2, y2 = crop_region
             w, h = x2 - x1, y2 - y1
             area = torch.tensor(w * h, device=im.device)
-            points_scale = np.array([[w, h]])  # w, h
+            points_scale = np.array([[iw, ih]])  # the crop is resized to the model input, so prompts live in that space
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/predict.py]: 📝 **LOW**: This scaling still breaks tensor sources with rectangular BCHW inputs because `_prepare_prompts()` below receives `self.batch[1][0].shape[:2]` as `(C, H)` rather than `(H, W)`. For example, a `(1, 3, 640, 960)` tensor gives `r=min(640/3, 960/640)=1.5`, so the new `[iw, ih]` grid produces points as large as `(1440, 960)`, outside the resized crop. Derive the source dimensions from the trailing axes for tensor inputs before prompt preparation.

File: ultralytics/models/sam/predict.py [hunk 3]
```diff
@@ -438,6 +441,13 @@ def generate(
             keep = torchvision.ops.nms(pred_bboxes, scores, crop_nms_thresh)
             pred_masks, pred_bboxes, pred_scores = pred_masks[keep], pred_bboxes[keep], pred_scores[keep]
 
+        if min_mask_region_area > 0:
+            h0, w0 = self.batch[1][0].shape[:2]  # HWC source; a tensor source is already imgsz x imgsz, so gain is 1
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/predict.py]: 📝 **LOW**: The original-image area conversion still assumes an HWC source. `LoadTensor` supplies BCHW data, so for a `(1, 3, 640, 960)` tensor `self.batch[1][0].shape[:2]` is `(3, 640)`, producing `gain=1.5` instead of `1`; a 100-pixel threshold is consequently treated as 225 model-space pixels and removes valid regions. Use `(H, W)` from the tensor's trailing dimensions while retaining `shape[:2]` for HWC inputs.

File: ultralytics/models/sam/predict.py [hunk 4]
```diff
@@ -438,6 +441,13 @@ def generate(
...
+            gain = min(ih / h0, iw / w0)  # masks are in letterboxed model space, the threshold is in original pixels
+            min_area = min_mask_region_area * gain * gain
+            pred_masks, keep = self.remove_small_regions(pred_masks, min_area, max(self.args.iou, crop_nms_thresh))
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/predict.py]: 📝 **LOW**: The confidence threshold remains inconsistent when this path is used through `Predictor.__call__`: `conf_thres` filters candidates in `generate()`, but `postprocess()` still filters `pred_scores` with `self.args.conf`. Thus `predictor(source=..., conf_thres=0.1, min_mask_region_area=100)` can discard masks scoring 0.1–0.25 after generation. Propagate the effective threshold into postprocessing or apply the final confidence filter in `generate()`.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 5

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since the previous review, the documentation and empty-input changes remain sound, but the two previously reported issues are still unresolved: tensor-source area scaling and the confidence-threshold mismatch. The revised crop scaling also exposes the same BCHW/HWC shape assumption for rectangular tensor inputs, causing out-of-bounds crop prompts. These are bounded LOW-severity correctness issues; no security or data-loss problems were found.

<details><summary>💬 Posted 3 inline comments</summary>

- 📝 **LOW** `ultralytics/models/sam/predict.py:388` This scaling still breaks tensor sources with rectangular BCHW inputs because `_prepare_prompts()` below receives `self.batch[1][0].shape[:2]` as `(C, H)` rather than `(H, W)`. For example, a `(1, 3, 640, 960)` tensor gives `r=min(640/3, 960/640)=1.5`, so the new `[iw, ih]` grid produces points as large as `(1440, 960)`, outside the resized crop. Derive the source dimensions from the trailing axes for tensor inputs before prompt preparation.
- 📝 **LOW** `ultralytics/models/sam/predict.py:445` The original-image area conversion still assumes an HWC source. `LoadTensor` supplies BCHW data, so for a `(1, 3, 640, 960)` tensor `self.batch[1][0].shape[:2]` is `(3, 640)`, producing `gain=1.5` instead of `1`; a 100-pixel threshold is consequently treated as 225 model-space pixels and removes valid regions. Use `(H, W)` from the tensor's trailing dimensions while retaining `shape[:2]` for HWC inputs.
- 📝 **LOW** `ultralytics/models/sam/predict.py:448` The confidence threshold remains inconsistent when this path is used through `Predictor.__call__`: `conf_thres` filters candidates in `generate()`, but `postprocess()` still filters `pred_scores` with `self.args.conf`. Thus `predictor(source=..., conf_thres=0.1, min_mask_region_area=100)` can discard masks scoring 0.1–0.25 after generation. Propagate the effective threshold into postprocessing or apply the final confidence filter in `generate()`.
</details>

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review 6

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 5, the BCHW/HWC source-shape handling has been centralized through `src_shape`, and the confidence-threshold mismatch has been resolved by filtering inside `generate()` before cleanup. The crop prompt scaling, area conversion, mask/score/box synchronization, empty-input handling, and documentation update look correct. No remaining concrete correctness, compatibility, security, or performance issues found. LGTM.

**Comment by UltralyticsAssistant** [comment]
🎉 PR [#26129](https://github.com/ultralytics/ultralytics/pull/26129) has been merged—thank you @raimbekovm!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This contribution brings together valuable SAM improvements: configurable small-region cleanup, corrected crop prompt scaling across resized inputs, consistent source-shape handling, and aligned confidence filtering in `generate()`, while preserving existing default behavior.
